src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java

```diff
@@ -413,115 +394,75 @@ private int toChars(long f, int e, FormattedFPDecimal fd) {
+    private int toChars1(byte[] str, int index, int h, int m, int l, int e) {
         /*
          * 0 < e <= 7: plain format without leading zeroes.
          * Left-to-right digits extraction:
          * algorithm 1 in [3], with b = 10, k = 8, n = 28.
          */
-        appendDigit(h);
+        putDigit(str, index++, h);
         int y = y(m);
         int t;
         int i = 1;
         for (; i < e; ++i) {
             t = 10 * y;
-            appendDigit(t >>> 28);
+            putDigit(str, index++, t >>> 28);
             y = t & MASK_28;
         }
-        append('.');
+        putChar(str, index++, '.');
         for (; i <= 8; ++i) {
             t = 10 * y;
-            appendDigit(t >>> 28);
+            putDigit(str, index++, t >>> 28);
             y = t & MASK_28;
         }
-        lowDigits(l);
-        return NON_SPECIAL;
+        return lowDigits(str, index, l);
     }
 
-    private int toChars2(int h, int m, int l, int e) {
+    private int toChars2(byte[] str, int index, int h, int m, int l, int e) {
         /* -3 < e <= 0: plain format with leading zeroes */
-        appendDigit(0);
-        append('.');
+        putCharsAt(str, index, '0', '.');
+        index += 2;
         for (; e < 0; ++e) {
-            appendDigit(0);
+            putDigit(str, index++, 0);
         }
-        appendDigit(h);
-        append8Digits(m);
-        lowDigits(l);
-        return NON_SPECIAL;
+        putDigit(str, index, h);
+        put8Digits(str, index + 1, m);
+        return lowDigits(str, index + 9, l);
     }
 
-    private int toChars3(int h, int m, int l, int e) {
+    private int toChars3(byte[] str, int index, int h, int m, int l, int e) {
         /* -3 >= e | e > 7: computerized scientific notation */
-        appendDigit(h);
-        append('.');
-        append8Digits(m);
-        lowDigits(l);
-        exponent(e - 1);
-        return NON_SPECIAL;
+        putCharsAt(str, index, '0' + h, '.');
+        put8Digits(str, index + 2, m);
+        index = lowDigits(str, index + 10, l);
+        return exponent(str, index, e - 1);
     }
 
-    private void lowDigits(int l) {
+    private int lowDigits(byte[] str, int index, int l) {
         if (l != 0) {
-            append8Digits(l);
-        }
-        removeTrailingZeroes();
-    }
-
-    private void append8Digits(int m) {
-        /*
-         * Left-to-right digits extraction:
-         * algorithm 1 in [3], with b = 10, k = 8, n = 28.
-         */
-        int y = y(m);
-        for (int i = 0; i < 8; ++i) {
-            int t = 10 * y;
-            appendDigit(t >>> 28);
-            y = t & MASK_28;
-        }
-    }
-
-    private void removeTrailingZeroes() {
-        while (bytes[index] == '0') {
-            --index;
-        }
-        /* ... but do not remove the one directly to the right of '.' */
-        if (bytes[index] == '.') {
-            ++index;
+            put8Digits(str, index, l);
+            index += 8;
```

Comment:
As discussed in `ToDecimal`, all `put...` methods should return the updated `index`. Then this code can be rewritten as
```suggestion
            index = put8Digits(str, index, l);
```
so that the caller doesn't need to know how many characters were added.